tests/test_generate_file.py

```diff
@@ -68,12 +68,18 @@ def test_generate_file_jsonify_filter(env) -> None:
 
 
 @pytest.mark.parametrize("length", (10, 40))
+@pytest.mark.parametrize("numbers", (True, False))
 @pytest.mark.parametrize("punctuation", (True, False))
-def test_generate_file_random_ascii_string(env, length, punctuation) -> None:
+def test_generate_file_random_ascii_string(env, length, numbers, punctuation) -> None:
     """Verify correct work of random_ascii_string extension on file generation."""
     infile = 'tests/files/{{cookiecutter.random_string_file}}.txt'
     data = {'random_string_file': 'cheese'}
-    context = {"cookiecutter": data, "length": length, "punctuation": punctuation}
+    context = {
+        "cookiecutter": data,
+        "length": length,
+        "numbers": numbers,
+        "punctuation": punctuation,
+    }
     generate.generate_file(project_dir=".", infile=infile, context=context, env=env)
     assert os.path.isfile('tests/files/cheese.txt')
     generated_text = Path('tests/files/cheese.txt').read_text()
```

Comment:
> This needs to confirm that numbers and/or punctuation are (or are not) present in the generated text...though technically, because it's random, it's possible that when `numbers` is True there might still not be numbers in the text.

Regarding the possibility that a string generated with numbers=True might not include any numbers (due to randomness): should the test only verify the presence or absence of numbers/punctuation based on the provided flags, or should the string generation function (random_ascii_string) guarantee that at least one number/punctuation character is included when the respective flags are set to True.

Reply:
That's a good question. We should probably mandate that at least one character from each requested class is present, but that may be technically difficult to enforce while still being random.

@ericof / @jensens, the question here is: "If numbers and/or punctuation are requested in a random string, should we guarantee that at least one number and/or punctuation mark is included?"

What are your thoughts?

If we do want to mandate that at least one number _must_ be present if requested, then we also need to make it clear that a call like

```jinja
{{ random_ascii_string(1, numbers=True, punctuation=True) }}
```

cannot guarantee that a character from each class is included (perhaps by raising an exception).

Reply:
> That's a good question. We should probably mandate that at least one character from each requested class is present, but that may be technically difficult to enforce while still being random.
> 
> @ericof / @jensens, the question here is: "If numbers and/or punctuation are requested in a random string, should we guarantee that at least one number and/or punctuation mark is included?"
> 
> What are your thoughts?
> 
> If we do want to mandate that at least one number _must_ be present if requested, then we also need to make it clear that a call like
> 
> ```
> {{ random_ascii_string(1, numbers=True, punctuation=True) }}
> ```
> 
> cannot guarantee that a character from each class is included (perhaps by raising an exception).

In my opinion, since the string generation is meant to be random, it shouldn’t be necessary to guarantee that a number or punctuation is always included when requested. This keeps the randomness aspect intact.

However, if we decide to enforce that at least one character from each class must be present, I can modify the random_ascii_string function to ensure this and add a check for short strings (like random_ascii_string(1, numbers=True, punctuation=True)), raising an exception if the length is too short.

Reply:
I'm on board for keeping the current semantics (`True` means "allowed", not "required). :+1: 

Could you add some test logic that confirms that if `numbers is False` then there are no numbers in the random string? Same for punctuation?